[TG-2922] Improve logging for byte flatten exceptions

thk123 opened the pull request:
Depends on: #2071 :white_check_mark: 

Submodule bump: https://github.com/diffblue/test-gen/pull/1739

A client has code that trips over this exception. However, we don't have access to the code so this change provides more information about the exception by re-throwing the exception back up the stack to record relevant information (like what kind of line is causing the problem). 

TODO:

- [x] Adapt existing unit test for the string utils into a catch test
- [x] Fix commit history
- [x] Add the original goto instruction corresponding to the symex step exception. 

Currently produces something like:

```
file TestClasses.java line 4 function java::Test1.f:(I)V: exception: Error converting guard for step
Souce GOTO statement: ((struct java::array[float] *)a1->data[j])->data[0] = 34.5f;
Step:
Thread 0 file TestClasses.java line 6 function java::Test1.f:(I)V bytecode-index 11
ASSIGNMENT (STATE)
dynamic_6_array#4 == (((struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object5 ? dynamic_object5#5.data == dynamic_6_array : ((struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object3 ? dynamic_object3#4.data == dynamic_6_array : invalid_object1#0.data == dynamic_6_array)) ? dynamic_6_array#3 WITH [0l:=34.5f] : dynamic_6_array#3)
Guard: (struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object5 ? dynamic_object5#5.data == dynamic_6_array : ((struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object3 ? dynamic_object3#4.data == dynamic_6_array : invalid_object1#0.data == dynamic_6_array)
 exception: Can't convert byte_extraction
  exception: cannot unpack array of non-const size
  array_type: array
    * size: symbol
        * type: signedbv
            * width: 32
            * #c_type: signed_int
        * identifier: java::Test1.f:(I)V::y!0@1#1
        * expression: symbol
            * type: signedbv
                * width: 32
                * #c_type: signed_int
            * identifier: java::Test1.f:(I)V::y
        * L0: 0
        * L1: 1
        * L2: 1
        * L1_object_identifier: java::Test1.f:(I)V::y!0@1
        * #SSA_symbol: 1
    * #dynamic: 1
    0: pointer
        * width: 64
        * #reference: 1
        0: struct
            * tag: java::array[float]
            * components: 
              0: 
                * type: struct
                    * incomplete_class: 1
                    * name: java::java.lang.Object
                    * tag: java.lang.Object
                    * base_name: java.lang.Object
                    * components: 
                      0: 
                        * type: string
                        * name: @class_identifier
                        * pretty_name: @class_identifier
                      1: 
                        * type: c_bool
                            * width: 8
                        * name: @lock
                        * pretty_name: @lock
                    * #class: 1
                * name: @java.lang.Object
                * base_name: @java.lang.Object
                * pretty_name: @java.lang.Object
              1: 
                * type: signedbv
                    * width: 32
                    * #c_type: signed_int
                * name: length
                * base_name: length
                * pretty_name: length
              2: 
                * type: pointer
                    * width: 64
                    * #reference: 1
                    0: floatbv
                        * width: 32
                        * f: 23
                * name: data
                * base_name: data
                * pretty_name: data
            * bases: 
              0: base
                  * type: symbol
                      * identifier: java::java.lang.Object
            * #class: 1
  max_bytes: +
    * type: unsignedbv
        * width: 64
        * #c_type: unsigned_long_int
    0: constant
        * type: unsignedbv
            * width: 64
            * #c_type: unsigned_long_int
        * value: 0000000000000000000000000000000000000000000000000000000000001000
    1: typecast
        * type: unsignedbv
            * width: 64
            * #c_type: unsigned_long_int
        0: pointer_offset
            * type: signedbv
                * width: 64
                * #c_type: signed_long_int
            0: member
                * type: pointer
                    * width: 64
                    * #reference: 1
                    0: pointer
                        * width: 64
                        0: empty
                * component_name: data
                0: symbol
                    * type: struct
                        * tag: java::array[reference]
                        * components: 
                          0: 
                            * type: struct
                                * incomplete_class: 1
                                * name: java::java.lang.Object
                                * tag: java.lang.Object
                                * base_name: java.lang.Object
                                * components: 
                                  0: 
                                    * type: string
                                    * name: @class_identifier
                                    * pretty_name: @class_identifier
                                  1: 
                                    * type: c_bool
                                        * width: 8
                                    * name: @lock
                                    * pretty_name: @lock
                                * #class: 1
                            * name: @java.lang.Object
                            * base_name: @java.lang.Object
                            * pretty_name: @java.lang.Object
                          1: 
                            * type: signedbv
                                * width: 32
                                * #c_type: signed_int
                            * name: length
                            * base_name: length
                            * pretty_name: length
                          2: 
                            * type: pointer
                                * width: 64
                                * #reference: 1
                                0: pointer
                                    * width: 64
                                    0: empty
                            * name: data
                            * base_name: data
                            * pretty_name: data
                        * bases: 
                          0: base
                              * type: symbol
                                  * identifier: java::java.lang.Object
                        * #class: 1
                    * identifier: symex_dynamic::dynamic_object1#3
                    * expression: symbol
                        * type: struct
                            * tag: java::array[reference]
                            * components: 
                              0: 
                                * type: struct
                                    * incomplete_class: 1
                                    * name: java::java.lang.Object
                                    * tag: java.lang.Object
                                    * base_name: java.lang.Object
                                    * components: 
                                      0: 
                                        * type: string
                                        * name: @class_identifier
                                        * pretty_name: @class_identifier
                                      1: 
                                        * type: c_bool
                                            * width: 8
                                        * name: @lock
                                        * pretty_name: @lock
                                    * #class: 1
                                * name: @java.lang.Object
                                * base_name: @java.lang.Object
                                * pretty_name: @java.lang.Object
                              1: 
                                * type: signedbv
                                    * width: 32
                                    * #c_type: signed_int
                                * name: length
                                * base_name: length
                                * pretty_name: length
                              2: 
                                * type: pointer
                                    * width: 64
                                    * #reference: 1
                                    0: pointer
                                        * width: 64
                                        0: empty
                                * name: data
                                * base_name: data
                                * pretty_name: data
                            * bases: 
                              0: base
                                  * type: symbol
                                      * identifier: java::java.lang.Object
                            * #class: 1
                        * identifier: symex_dynamic::dynamic_object1
                    * L2: 3
                    * L1_object_identifier: symex_dynamic::dynamic_object1
                    * #SSA_symbol: 1
```

Commit history still needs tidying up and I found some confusing behaviour in the string utils so want to finish off the unit tests for that.

Commit:
Introduce nested exception printing

Commit:
Convert flatten_byte_extract to use structured exceptions

Store error in string to avoid dangling pointers

tautschnig commented:
I suppose that's to an extent the result of a discussion that @svorenova and I had: it is really surprising (wrong?) that Java code would generate byte_extract expressions. And actually I don't really see why this needs to be related to inaccessible client code for @svorenova has a very small example that produced this as well?!

tautschnig commented:
More notes:
1. As said, the fact that Java code generates byte_extract seems broken to me. I may be wrong, but I think that's what should be fixed in the first instance.
2. This looks like pretty incredible machinery, and surely that's useful for debugging (notably C code will produce those kinds of byte_extract expressions).
3. Despite this, it's disaster recovery rather than preventing disaster. Check with @chrisr-diffblue and @danpoe on work to do the latter: such a byte_extract expression should never even hit the back-end. That's closely related to #1880.

majakusber commented:
@tautschnig You are correct. Since the bug report came from a client, we are approaching this in two steps - first extending the logging for crashes of this type and second fixing the bug (by avoiding the byte extract instructions as we discussed). This is a PR for the first step, I will be preparing a PR for the second step. Extended logging should allow the client to provide us more information in the future if they encounter this crash (it took us considerable amount of time to dig out what was the actual problem).

tautschnig commented:
> (it took us considerable amount of time to dig out what was the actual problem)

You are not alone in this. @chrisr-diffblue will be able to tell you more.

Edit: that's why I insist that 3. from my above comment is important.

thk123 commented:
@tautschnig That is mostly want I wanted to hear - that this is a reasonable approach for providing more info for these kind of errors, even when we fix the underlying problem. In this case we were able to match it up to an existing bug where we did have the source, but in general would have been hard to work out the problem without having a symmetric error in our own examples. 

I suppose there will be other errors arising from the symex phase where knowing where in the symex has been got to would be useful. This machinery will make these easier to add in in the future. 

Also thanks for the hint that @chrisr-diffblue is looking at something similar!

Commit:
Introduce throwing bv_conversion expection

Remove suprisous namespace from bv_conversion_exceptions

Commit:
Introduce throwing a guard conversion exception

Throw the unwrapped exception as a string

Commit:
Provide the original goto statement in the error

This required the symex target_equation having the namespace to resolve
names

thk123 commented:
I've updated this to also print out the original goto statement that the symex has failed to convert:

```
Source GOTO statement: ((struct java::array[float] *)a1->data[j])->data[0] = 34.5f;
```

This is probably the most important piece of information (in this case I believe clearly pointing towards writing to a multi-dimensional array. 

To do this I had to thread a `ns` into it, only one level deeper but I think it is much clearer with this extra info right at the top.

tautschnig commented:
> I've updated this to also print out the original goto statement that the symex has failed to convert: [...]

Very useful indeed! And then I keep wondering why this introduces a `byte_extract`...

majakusber reviewed (approved):
Looks good and useful. The CI failures look like minor code issues. Please consider squashing the commits for introducing/moving/adjusting the unwrap code.

thk123 commented:
So I split out the changes for the split_string (along with unit tests) into #2071 which this PR now depends on. Real diff for this commit (which remains changed from previous reviews): https://github.com/diffblue/cbmc/pull/2046/files/c00350b6d508ec40608a300db93d0288bd298304..705ee6592521f0682063273521a351c3244c0bf4 

Need to add the new file to the Makefile and will squash the commits as suggested by @svorenova

thk123 commented:
@svorenova Squashed as suggested. Also fixed the errors in Makefiles

Commit:
Disable nested exception printing for Windows

Due to bug in the VS2013 C++ compiler, using std::rethrow_if_nested or
std::nested_exception is not supported. This disables trying to unwrap the
exception and just prints a warning saying the nested exceptionc couldn't be
printed.

Don't use noexcept directly, pull both part of the nested exception into
a separate file to handle discrepancies.

Commit:
Test demonstrating logging with clause for dealing with Windows

thk123 commented:
Created a version that should work on Windows: https://github.com/diffblue/cbmc/pull/2046/commits/9a68db323002edb9ca13e97dc3e3b32d99e630d6

Also added a test so we actually check the logging works (fix for the bug to come)

tautschnig commented:
Since the "do not merge" label has been removed I've got to ask: my thinking was that this was an experimental feature. Please do state if that's not the case and there is an actual intent to get this merged.

thk123 commented:
@tautschnig indeed the intention is to merge this. The goal here is to be able to provide more useful error messages for this kind of error in the future. That naturally requires the code to be in releases.

thk123 commented:
Windows doesn't even appear to have the `std::nested_exception` so there is no way to unroll the exception. Log a warning saying you are not getting the full picture.

thk123 commented:
Should have seen that coming - throw_with_nested also not supported by Windows...

chrisr-diffblue commented:
Thanks @tautschnig and @thk123 for poking me about this issue. I've just had a face-to-face chat with  @thk123 about this. As @tautschnig points out, this is indeed very much a disaster recovery 'error handling of last resort' patch, but I would argue that it's still worth adding as a developer debugging aid. I don't think this is intended to be a user facing error message, and definitely we should/will be working towards ensuring that the user never hits debugging code like this where we can. @tautschnig - would you be happy to let this PR go through if there was also an appropriate issue raised to cover the 'disaster prevention' side of your concerns too?

tautschnig commented:
I have stated it elsewhere, but for completeness:
>  @tautschnig - would you be happy to let this PR go through if there was also an appropriate issue raised to cover the 'disaster prevention' side of your concerns too?

I will comment on bits that I think should be changed, but certainly I don't intend to block this PR in any way.

tautschnig commented on src/solvers/flattening/bv_conversion_exceptions.h:
```diff
@@ -0,0 +1,36 @@
+#include "flatten_byte_extract_exceptions.h"
+#include "flatten_byte_operators.h"
+#include <util/endianness_map.h>
+#include <util/arith_tools.h>
```
Nit picking: can the includes be grouped by directory and be lexicographically sorted?

tautschnig commented on src/solvers/flattening/boolbv.cpp:
```diff
@@ -152,6 +152,11 @@ bvt boolbvt::conversion_failed(const exprt &expr)
+/// \param expr: TODO
+/// \return TODO
+/// \throws bitvector_conversion_exceptiont raised if converting byte_extraction
+/// goes wrong. TODO: extend for other types of conversion exception).
```
Proposed documentation: Converts an expression into its gate-level representation and returns a vector of literals corresponding to the outputs of the Boolean circuit.

tautschnig commented on src/goto-symex/symex_target_equation.h:
```diff
@@ -320,6 +323,9 @@ class symex_target_equationt:public symex_targett
   void merge_ireps(SSA_stept &SSA_step);
+
+private:
+  const namespacet &ns;
```
This has been removed in e8eec2c8a8 so I'm a bit hesitant seeing it re-introduced. @karkhaz would you recall why it had to be removed?

karkhaz replied:
There's a possibility that the namespace that this reference points to would go out of scope if we save this `symex_target_equationt` and then resume it later. For some classes (e.g. `goto_symex_statet`), my solution to this was to [make the copy constructor private](https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_symex_state.h#L380) and [force clients to pass an equation explicitly to the copy constructor](https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_symex_state.h#L38).

I think the reason we decided not to do this for the `ns` member of `symex_target_equationt` was that there were no users of `ns` in the class (or maybe just the `output` method?) so we just decided to delete the member. If you would like to have a namespace reference again, I would suggest that you follow the approach I used in those links I posted above.

thk123 replied:
I suppose if I use `format` I can drop this back out anyway

tautschnig replied:
That would be my preferred solution, yes.

thk123 replied:
Means we can't output the `SSA_stept` - so I'd prefer to keep it unless there is an equivalent way to work around that?

tautschnig replied:
I'd see two options: make the `output`  function use `format` instead of `from_expr` and remove the `namespacet` parameter or introduce an `output_generic` function that does this while leaving the existing `output` function intact. In a way, such back-end code shouldn't rely on language specifics anyway and thus shouldn't be doing language-specific output either.

thk123 replied:
@tautschnig Created an overload without the `ns` and deprecated the old one. https://github.com/diffblue/cbmc/pull/2046/commits/069eeb1792f491a4129852ae0e947b629b4a0ee6

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -417,7 +431,16 @@ void symex_target_equationt::convert_guards(
+      }
+      catch(const bitvector_conversion_exceptiont &conversion_exception)
+      {
+        util_throw_with_nested(guard_conversion_exceptiont(step, ns));
```
I believe the same try/catch is necessary for all convert_* routines.

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -368,7 +372,17 @@ void symex_target_equationt::constraint(
+  {
+    // unwrap the except and throw like normal
+    const std::string full_error = unwrap_exception(guard_conversion_exception);
+    throw full_error;
```
Don't restrict this to guards.

tautschnig commented on src/goto-symex/equation_conversion_exceptions.h:
```diff
@@ -0,0 +1,53 @@
+    std::ostringstream error_msg;
+    error_msg << runtime_error::what();
+    error_msg << "\nSource GOTO statement: "
+              << from_expr(ns, "java", step.source.pc->code);
```
Use `format(...)` from `util/format_expr.h` to avoid 1) the need for a namespace and 2) language-specifics when you don't actually have information about what language it is.

tautschnig commented on regression/cbmc-java/dynamic-multi-dimensional-array/test.desc:
```diff
@@ -0,0 +1,9 @@
+TestClass.class
+--function TestClass.f --cover location --unwind 2
+Source GOTO statement: .*
+(^ exception: Can't convert byte_extraction|warning: Nested exception printing not supported on Windows)
```
The output doesn't actually include "warning: ", hence this currently fails on Windows/AppVeyor.

Commit:
Add documentation to convert_bitvector

Commit:
Introduce exceptions for all conversion steps.

thk123 commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -416,7 +415,16 @@ void symex_target_equationt::convert_decls(
+      {
+        util_throw_with_nested(
+          equation_conversion_exceptiont(
+            "Error converting guard for step", step, ns));
```
Copy paste error! isn't converting guard step

thk123 commented:
@tautschnig I've added a (very) basic version for the other conversion steps. https://github.com/diffblue/cbmc/pull/2046/commits/f7fd0a123c63a8628f4cf84036e91ae9fb060f54

I don't have the bandwidth to expand it at the moment, but I think this is at least a mild improvement for all the cases. 

I've messed up the ordering of includes in a couple of other commits, and copy pasted nonsense which I'll fix now - fixed :white_check_mark: 

Could you let me know if you're happy with:

- Keeping the `ns` for the sake of printing the `SSA_stept` 
- The extra exception stuff for the other conversion steps (modulo the message being correct!)

thk123 commented on src/util/unwrap_nested_exception.cpp:
```diff
@@ -0,0 +1,56 @@
+
+#include <util/invariant.h>
+#include <util/string_utils.h>
+#include <sstream>
```
Include in wrong place

thk123 commented on src/util/unwrap_nested_exception.h:
```diff
@@ -0,0 +1,17 @@
+#define CPROVER_UTIL_UNWRAP_NESTED_EXCEPTION_H
+
+#include <string>
+#include <exception>
```
Includes not correctly ordered

thk123 commented on src/solvers/flattening/flatten_byte_extract_exceptions.h:
```diff
@@ -0,0 +1,147 @@
+
+#include <sstream>
+#include <iostream>
+#include <util/simplify_expr.h>
```
Not ordered correctly

thk123 commented on src/solvers/flattening/flatten_byte_operators.cpp:
```diff
@@ -18,13 +18,17 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/simplify_expr.h>
 
 #include "flatten_byte_operators.h"
+#include "flatten_byte_extract_exceptions.h"
```
Not ordered correctly

thk123 commented on src/solvers/flattening/boolbv_byte_extract.cpp:
```diff
@@ -16,6 +16,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/endianness_map.h>
 
 #include "flatten_byte_operators.h"
+#include "flatten_byte_extract_exceptions.h"
```
Not ordered correctly

thk123 commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -12,13 +12,16 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <solvers/prop/prop_conv.h>
 #include <solvers/prop/prop.h>
 #include <solvers/prop/literal_expr.h>
+#include <solvers/flattening/bv_conversion_exceptions.h>
```
Not ordered correctly

thk123 commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -12,13 +12,16 @@ Author: Daniel Kroening, kroening@kroening.com
+#include <solvers/flattening/bv_conversion_exceptions.h>
 
 #include "goto_symex_state.h"
+#include "equation_conversion_exceptions.h"
```
Not ordered correctly

thk123 reviewed (commented):
Various include orders

Commit:
Use format rather than from_expr for output

thk123 commented:
@tautschnig Due to code owner permissions, once you're happy (As implied by the :+1:), could you approve.

(@svorenova Might be worth a re-review of the final 5 commits as they are new since your review)

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -12,12 +12,17 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <solvers/prop/literal_expr.h>
+#include <solvers/prop/prop.h>
+#include <solvers/prop/prop_conv.h>
+#include <util/format.h>
```
1. Place all `util/` includes in one group.
2. You might add a comment with the `language_util.h` include that it can be removed once the now-deprecated function is gone.

tautschnig commented on src/goto-symex/symex_target_equation.h:
```diff
@@ -32,7 +32,6 @@ class prop_convt;
 class symex_target_equationt:public symex_targett
 {
 public:
-  symex_target_equationt() = default;
```
I'm not currently reviewing this commit-by-commit so please excuse if my comment is wrong: this should go in a commit of its own. I agree that it can safely be removed, but this shouldn't be done as part of some random changes.

thk123 replied:
So this happened as I went from `= default` -> one that takes a namespace -> `no constructor`. Will update the last commit to go back to `= default` and while I'm in the area append an extra commit to remove it. I think this is the most logical commit history

tautschnig replied:
Brilliant, thank you very much!

tautschnig commented on src/solvers/flattening/boolbv.cpp:
```diff
@@ -152,6 +152,13 @@ bvt boolbvt::conversion_failed(const exprt &expr)
+/// \return A vector of literals corresponding to the outputs of the Boolean
+///   circuit
+/// \throws bitvector_conversion_exceptiont raised if converting byte_extraction
+/// goes wrong. TODO: extend for other types of conversion exception).
```
Please create an issue for this "TODO" and include its id in the comment text.

thk123 replied:
#2103

tautschnig commented on src/goto-symex/equation_conversion_exceptions.h:
```diff
@@ -0,0 +1,59 @@
+
+#include <ostream>
+
+#include <util/namespace.h>
```
Not needed/used.

tautschnig commented on src/goto-symex/equation_conversion_exceptions.h:
```diff
@@ -0,0 +1,59 @@
+
+#include <util/namespace.h>
+
+#include <langapi/language_util.h>
```
Not needed/used.

tautschnig commented on src/solvers/flattening/bv_conversion_exceptions.h:
```diff
@@ -0,0 +1,37 @@
+#define CPROVER_SOLVERS_FLATTENING_BV_CONVERSION_EXCEPTIONS_H
+
+#include <util/arith_tools.h>
+#include <util/endianness_map.h>
```
Neither of those seem to be needed/used.

tautschnig commented on src/solvers/flattening/bv_conversion_exceptions.h:
```diff
@@ -0,0 +1,37 @@
+#include "boolbv.h"
+#include "bv_conversion_exceptions.h"
+#include "flatten_byte_extract_exceptions.h"
+#include "flatten_byte_operators.h"
```
I think only `bv_conversion_exceptions.h` is the only required one.

thk123 replied:
These includes just seem like nonsense - `bv_conversion_exceptions.h` must not be needed since that's the file we're in! Replaced with what appear to me to be correct

tautschnig commented on src/solvers/flattening/flatten_byte_extract_exceptions.h:
```diff
@@ -0,0 +1,147 @@
+#ifndef CPROVER_SOLVERS_FLATTENING_FLATTEN_BYTE_EXTRACT_EXCEPTIONS_H
+#define CPROVER_SOLVERS_FLATTENING_FLATTEN_BYTE_EXTRACT_EXCEPTIONS_H
+
+#include <iostream>
```
Seems unnecessary.

tautschnig commented on src/solvers/flattening/flatten_byte_extract_exceptions.h:
```diff
@@ -0,0 +1,147 @@
+#include <util/namespace.h>
+#include <util/pointer_offset_size.h>
+#include <util/simplify_expr.h>
+#include <util/std_expr.h>
```
I think most of the above are unnecessary.

tautschnig commented on src/util/format.h:
```diff
@@ -10,6 +10,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <iosfwd>
+#include "format_expr.h"
+#include "format_type.h"
```
No, please don't. Why would this be necessary?

thk123 replied:
Need the declaration where used - I thought it made more sense to include them here so people can use `format.h` without further includes, but happy to devolve this to use sites.

tautschnig replied:
Yes, code should use `format_expr.h` or `format_type.h`. If `format.h` were to be used directly then this should be done consistently across the code base.

tautschnig commented on src/util/unwrap_nested_exception.cpp:
```diff
@@ -0,0 +1,57 @@
+
+#include <sstream>
+#include <util/invariant.h>
+#include <util/string_utils.h>
```
Use `"..."` within the folder instead of angle brackets.

tautschnig commented on src/util/unwrap_nested_exception.cpp:
```diff
@@ -0,0 +1,57 @@
+  catch(const std::exception &e)
+  {
+    std::string nested_message = unwrap_exception(e, level + 1);
+    // Some exception messages already end in  a new line (e.g. as they have
```
Nit pick: two spaces between `in` and `a`.

tautschnig commented on src/util/unwrap_nested_exception.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "throw_with_nested.h"
+
+#include <exception>
+#include <string>
```
The header file already includes those two.

majakusber commented on src/util/throw_with_nested.h:
```diff
@@ -0,0 +1,59 @@
+
+#ifdef _MSC_VER
+#include <stdexcept>
+// TODO(tkiley): Nested exception logging not supported on windowsdue to bug
```
Typo: windowsdue to bug -> Windows due to a bug

thk123 replied:
#2104

majakusber reviewed (approved):
Looks good, only one typo and please make sure to put the TODO ticket number in the code too.

Commit:
Remove redundant default constructor

Commit:
Reformatting touched output function

thk123 commented:
All comments addressed I believe and TG pointer bump passing.

tautschnig commented on regression/cbmc-java/dynamic-multi-dimensional-array/test.desc:
```diff
@@ -0,0 +1,9 @@
+^EXIT=6$
+--
+--
+This is a bug - this validates the error output.
```
Nit pick: what does "This" refer to? The fact that this exception fires is some sort of bug, but the test case as set up now isn't. How about "The exception thrown in this test is the symptom of a bug; the purpose of this test is the validate the output of that exception."

tautschnig commented on src/goto-symex/equation_conversion_exceptions.h:
```diff
@@ -0,0 +1,47 @@
+
+#include <ostream>
+
+#include <util/format.h>
```
`format_expr.h`

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -11,13 +11,21 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "symex_target_equation.h"
 
+#include <util/format.h>
```
Unnecessary.

tautschnig reviewed (approved):
Looks good modulo the remaining nit picks.
